### ISSUE TITLE
Clink crashes with cmd prompts (e.g. 'set /p a=')

### ISSUE THREAD

**Issue by tgice** [chrisant996/clink#159]
Firstly, thanks for taking over the Clink project... I love having this in Windows cmd, it's so much of an upgrade over the clunky old shell.

Second, I've been using the original version for a couple of years (from mridgers) and started having some regular problem with it crashing (or something similarly disruptive) out of the blue and started looking for solutions and saw a post from you replying to another user's issue on the mridgers project.

I started using your project a couple of days ago, and am currently running your latest, 1.2.30.0ddff3.

I started experiencing a crash today on my main workstation, first with the previous version (1.2.29 I think) and the upgrade to the latest did not solve it.  I'll also admit that I did a fresh install of 1.2.30 on another Win10 machine and for some reason do not see the crash there, so it must be something odd about my primary workstation that's causing the issue.  I was hoping you might help me diagnose the issue.

Steps to duplicate the crash:

```
cd %temp%
echo>t
copy t t2
copy t t2
```

(it seems to happen because of the prompt asking to confirm whether to overwrite t2 after the second copy executes) -- then you get the "CLINK'S CRASHED"  message, and it says that's it's writing a core dump but unfortunately that file seems to be 0 bytes in:

%LOCALAPPDATA%\clink

I also thought it could've had something to do with clink's history & settings files so I uninstalled, and had the installer remove the scripts/history etc., and when that didn't seem to wipe the above LOCALAPPDATA dir, I manually moved it and re-installed.

I get the same behavior.  Any insight into what could cause this, what else I could do to get some more help tracking the issue or other packages to attempt to uninstall/reinstall to help with this?

I have a lot of software installed on this PC, and I'm sure something may be interacting in a bad way, but don't know where to begin.

Thanks,

John

**Comment by tgice** [comment]
Quick update: I realized that the crash I started experiencing in the last couple of weeks with the last of mridgers' version also seems to happen with your latest version, so it's nothing new apparently in clink, it's just something new on my PC.

A simpler reproduction of the error is simply a set that causes a user prompt:

`set /p a=`

**Comment by chrisant996** [comment]
> Quick update: I realized that the crash I started experiencing in the last couple of weeks with the last of mridgers' version also seems to happen with your latest version, so it's nothing new apparently in clink, it's just something new on my PC.
> 
> A simpler reproduction of the error is simply a set that causes a user prompt:
> 
> `set /p a=`

All known `set /p` issues were fixed a while back in the new Clink.  I'm not able to reproduce the crash.

Is there any additional info you can share that might help me get the crash to happen on my system, so that I can debug it?

My understanding is the steps are:
1. Install fresh clink.
2. Verify there are no *.lua files in the scripts directories listed by `clink info` (other than maybe a placeholder clink.lua file next to the clink executable files).
3. Start plain cmd.exe without any custom terminal programs.
4. `cd %temp%`
5. `echo>t`
6. `copy t t2`
7. `copy t t2`
8. And then it crashes immediately -- or does it crash after entering something at the confirmation prompt?

A crash dump would be ideal, but it's strange that the `%LOCALAPPDATA%\clink\clink.dmp` file is 0 bytes.  I'm not sure what to suggest.

**Comment by chrisant996** [comment]
Are you using an alternative terminal program, such as ConEmu or Windows Terminal or etc?  Does the issue occur in the default OS terminal, or only in certain alternative terminal programs?

**Comment by tgice** [comment]
Hi Chris,

I've rebooted at least once and this problem persists.  I did have ConEmu installed (from a months-ago experiment) but I was not using.  I've since uninstalled it and still see the same problem, easily reproduced with:

`set /p a=`

Still 0 byte clink.dmp file.

I'm only running this in the default Command Prompt (cmd.exe) launched from the Start Menu, and again I can't figure out what could've changed in the past couple of weeks to cause this (because I didn't start doing something new with it to run into this prompt situation that crashes Clink -- I'd always been doing that same action and it used to not crash).

I'm running:

Label|Value
---|---
Bits|64
Edition|Windows 10 Pro
Version|21H1
Installed on|6/‎23/‎2020
OS build|19043.1165
Experience|Windows Feature Experience Pack 120.2212.3530.0

I've got one optional update I've not yet installed that I doubt would fix this:

2021-08 Cumulative Update Preview for Windows 10 Version 21H1 for x64-based Systems (KB5005101)

Everything else is up-to-date I think.

Any other ideas of things to check or try?  Thanks for your time.

**Comment by chrisant996** [comment]
First, let's see if we can figure out why the dump file appears to be empty.

The screen should say something like this:

```
!!! CLINK'S CRASHED!
!!! v1.2.31.d8deec (x64)
!!!
!!! Writing core dump
!!! C:\Users\chrisant\AppData\Local\clink\clink.dmp
!!! ...ok
!!!
!!! Clink: 0x00007FFC3C190000
!!! Host: 0x00007FF7CD750000
!!! Guru: 0xc0000005 addr=0x0000000000000000 (8)
!!!
!!! Backtrace:
```

Below the dump file name:
- Does your screen say `...ok`?
- Or does it say `...failed`?

**Comment by tgice** [comment]
I've never seen it get past the %LOCALDATA%\clink\clink.dmp line.

It never says `...ok` or `...failed` -- but I will point out that the window disappears so quickly that I couldn't really see much anyway, but last time I happened, I carefully stared right at the bottom of the window looking for anything else and could not see anything as the window closed.

Does this suggest to you that clink is having a problem writing the dmp file?  Like a permissions/access problem or something?  

Anything else I could try to maybe get the window to stick around longer (or until I press a key) to make sure clink isn't putting something else out that I'm not seeing?

p.s. As expected, when I upgraded the non-crashing PC to the identical build level as the crashing one, the first one *still* does not crash.

**Comment by chrisant996** [comment]
> I've never seen it get past the %LOCALDATA%\clink\clink.dmp line.
> 
> It never says `...ok` or `...failed` -- but I will point out that the window disappears so quickly that I couldn't really see much anyway

Normally the window does not close when a crash happens.  It shows information about the crash and where to find the crash dump, and it waits until you press a key.  It seems something deeper is going wrong, and causing enough instability that maybe it crashes again while trying to handle the first crash.

Can you please attach:
1. The output from `clink info`.
2. The `clink_settings` file.
3. The `clink.log` file.

If privacy is a concern, you can put them in a password encrypted .zip file and send me the password privately at the email address listed in my github profile.

Also, if you are willing and able to, can you try turning off anti-virus protection before starting Clink, and then turn the protection on again after trying to reproduce the crash?  It's possible for anti-virus software to interfere in unexpected ways, and sometimes turning off anti-virus protection clears up mystery issues (in which case turn the protection on again and report the false positive to the anti-virus product support).  However, while I know that I uploaded Clink without any malware, I don't know whether malware already exists on your machine, and so I don't know how safe turning off anti-virus protection on your machine is -- only you can make that decision.

**Comment by chrisant996** [comment]
Also -- if you happen to be familiar with using a debugger, then you could try this:
1. Attach a debugger to CMD.exe
2. Reproduce the crash
3. The debugger will take over -- at that point it would be possible to tell the debugger to save a crash dump, and that should work.

**Comment by tgice** [comment]
> Also, if you are willing and able to, can you try turning off anti-virus protection before starting Clink, and then turn the protection on again after trying to reproduce the crash? It's possible for anti-virus software to interfere in unexpected ways, and sometimes turning off anti-virus protection clears up mystery issues (in which case turn the protection on again and report the false positive to the anti-virus product support). However, while I know that I uploaded Clink without any malware,

Hey, you know what, that just triggered me to remember what may've caused all of this -- sometime in the past couple of weeks, I did have my corporate PC (the one with the crashing problem) migrated from an older Sophos anti-virus solution to a new TrendMicro one.  This new one has been causing me a few other problems launching software (including your version of Clink, especially when I was trying to run it from the portable install - once I switched to the installer, it just bugged me at first but then allowed the install).

So .... *that* probably is the issue.  But since I'm not involved with managing TrendMicro, I'm going to have to submit a ticket and wait for them to get back to me.  I will follow up with the results of pausing or removing TrendMicro.

Thanks!

**Comment by chrisant996** [comment]
> sometime in the past couple of weeks, I did have my corporate PC (the one with the crashing problem) migrated from an older Sophos anti-virus solution to a new TrendMicro one. This new one has been causing me a few other problems launching software (including your version of Clink, especially when I was trying to run it from the portable install - once I switched to the installer, it just bugged me at first but then allowed the install).

Ah, that makes a lot of sense.

> So .... _that_ probably is the issue. But since I'm not involved with managing TrendMicro, I'm going to have to submit a ticket and wait for them to get back to me. I will follow up with the results of pausing or removing TrendMicro.

How inconvenient!  I'm sorry you have to deal with that.

**Do you have experience with using a debugger?**  If so then you could attach a debugger to cmd.exe, trigger the crash, and then use the debugger to save the crash dump.  That should work even if a malware protection suite is interfering.

**Otherwise,** I will close the issue until additional information is available.  However, if at some point you're able to successfully collect a crash dump file, then please reopen this issue and share whatever new information you've been able to collect.

**Comment by tgice** [comment]
Thanks for the Time Travel Trace advice.  I spent a little time trying to use the (older?) version of WinDbg (6.3.9600...) that was already installed on my Win 10 and was able to attach to the process, but didn't seem to figure out how to capture a run file.  I will see whether I can get the WinDbg Preview installed from the MS Store tomorrow and see if I can get you that run file before I'm able to workaround the issue by disabling/removing/etc TrendMicro.

Thanks for your followup!

**Comment by chrisant996** [comment]
> I spent a little time trying to use the (older?) version of WinDbg (6.3.9600...) that was already installed on my Win 10

Oh, you have WinDbg installed already -- perfect!

These steps should successfully capture a crash dump:
1. Start a new cmd.exe with Clink.
2. Run `clink info` and note the "session" number it reports.
3. Start WinDbg and attach to the cmd.exe process.
  - Click "File" in the menu ribbon, then click "Attach to process".
  - Click the "Process" column header to sort the list.
  - Find the cmd.exe process whose "PID" column matches the "session" number from step 2.
  - Click that one.
  - Click the "Attach" button in the lower right corner of the WinDbg window.
4. Do the steps that make Clink crash.
5. The crash should be caught by WinDbg.
6. Find the "Command" window in WinDbg.
7. In the edit box at the bottom of the "Command" window enter `.dump /mA c:\temp\clink.dmp` (or use an appropriate folder if `c:\temp` doesn't exist or isn't usable).

Then reply here and attach the `c:\temp\clink.dmp` in the reply.

Thanks!

**Comment by tgice** [comment]
Hi, I installed the WinDbg preview from the Microsoft Store, just in case that makes any difference and ran the dump there.

One thing to mention: as soon as I attach to the cmd process running Clink, I get a message like this:

```

Microsoft (R) Windows Debugger Version 10.0.22415.1003 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.

*** wait with pending attach
Symbol search path is: srv*
Executable search path is: 
ModLoad: 00007ff6`19990000 00007ff6`199f7000   C:\Windows\System32\cmd.exe
ModLoad: 00007ffb`02170000 00007ffb`02365000   C:\WINDOWS\SYSTEM32\ntdll.dll
ModLoad: 00007ffb`01210000 00007ffb`012cd000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007ffa`ffe20000 00007ffb`000e9000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007ffb`00290000 00007ffb`0032e000   C:\WINDOWS\System32\msvcrt.dll
ModLoad: 00007ffb`007b0000 00007ffb`00b05000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007ffa`ffd20000 00007ffa`ffe20000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00007ffb`01010000 00007ffb`0113a000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00007ffa`d96e0000 00007ffa`d98e2000   C:\WINDOWS\system32\tmumh\20019\AddOn\8.53.0.1118\TmUmEvt64.dll
ModLoad: 00007ffb`01e30000 00007ffb`01e38000   C:\WINDOWS\System32\PSAPI.DLL
ModLoad: 00007ffb`01c90000 00007ffb`01e30000   C:\WINDOWS\System32\USER32.dll
ModLoad: 00007ffa`ff9f0000 00007ffa`ffa12000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007ffb`00230000 00007ffb`0025a000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007ffa`ffa20000 00007ffa`ffb2b000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007ffa`ff950000 00007ffa`ff9ed000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00007ffb`012d0000 00007ffb`0137c000   C:\WINDOWS\System32\ADVAPI32.dll
ModLoad: 00007ffb`00b10000 00007ffb`00bab000   C:\WINDOWS\System32\sechost.dll
ModLoad: 00007ffb`00260000 00007ffb`00290000   C:\WINDOWS\System32\IMM32.DLL
ModLoad: 00000000`53080000 00000000`531ef000   C:\WINDOWS\system32\tmumh\20019\TmMon\2.8.0.1051\tmmon64.dll
ModLoad: 00007ffa`c8c90000 00007ffa`c8c9c000   C:\WINDOWS\SYSTEM32\cmdext.dll
ModLoad: 00007ffa`c3150000 00007ffa`c32c9000   C:\Users\jlawler\AppData\Local\Temp\clink\dll_cache\1.2.30.0ddffe_20ddf105\clink_dll_x64.dll
ModLoad: 00007ffb`00db0000 00007ffb`00e05000   C:\WINDOWS\System32\SHLWAPI.dll
ModLoad: 00007ffb`014f0000 00007ffb`01c2f000   C:\WINDOWS\System32\SHELL32.dll
ModLoad: 00007ffa`f6c80000 00007ffa`f6c8a000   C:\Windows\System32\VERSION.dll
ModLoad: 00007ffa`edda0000 00007ffa`edf84000   C:\Windows\System32\dbghelp.dll
ModLoad: 00007ffa`ed300000 00007ffa`ed32c000   C:\Windows\System32\dbgcore.DLL
ModLoad: 00007ffa`fd880000 00007ffa`fe010000   C:\WINDOWS\SYSTEM32\windows.storage.dll
ModLoad: 00007ffa`ff150000 00007ffa`ff17c000   C:\Windows\System32\Wldp.dll
ModLoad: 00007ffb`00f60000 00007ffb`0100e000   C:\WINDOWS\System32\SHCORE.dll
(63a4.53d8): Break instruction exception - code 80000003 (first chance)
ntdll!DbgBreakPoint:
00007ffb`02210910 cc              int     3
```

So I first have to click 'Go' to continue so I can actually enter commands into the cmd window.  Once I do that, I can enter the `set /p a=` that I know will trigger the issue, and then I see something like the above, with this text at the bottom:

```
(63a4.53d8): Break instruction exception - code 80000003 (first chance)
ntdll!DbgBreakPoint:
00007ffb`02210910 cc              int     3
0:003> g
(63a4.5a5c): Stack overflow - code c00000fd (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
KERNELBASE!VirtualQueryEx+0x19:
00007ffa`ffe8ae39 48ff15504e1500  call    qword ptr [KERNELBASE!_imp_NtQueryVirtualMemory (00007ffa`fffdfc90)] ds:00007ffa`fffdfc90={ntdll!NtQueryVirtualMemory (00007ffb`0220d270)}
```

I will email you a zip with the dmp file.

**Comment by chrisant996** [comment]
Oh! Even just knowing it's a stack overflow is very useful, and I look forward to taking a deeper look at the dmp file. Excellent. 😎

**Comment by chrisant996** [comment]
I received the crash dump, thanks!  Can you also attach the `clink.log` file?  (Use `clink info` to find it.)

Something strange happened during Clink's initialization, and hopefully the log file may include more info.  However, it kind of looks as though the antivirus suite might be blocking Clink from checking whether Clink has already been injected into the process, but doesn't block Clink from injecting.  Thus resulting in Clink injecting itself multiple times.

If that were to happen, it would create a problem exactly like this one.  The issue is that Clink hooks the ReadConsoleW API so that it can take over the input responsibilities.  But in some cases Clink needs to call the original ReadConsoleW and let the OS handle input responsibilities (like at the prompt whether to overwrite a file, or a `set /p` prompt).  When Clink calls "the original function", lo and behold it is Clink's own function!  That would of course happen if Clink injected twice -- the first time it will save the original function and then add its own -- the second time it will save the "original" function again, but uh oh it's already been hooked by Clink, so it's not actually the original function anymore.  So when Clink's read_console hook function tries to call the original function, it gets misdirected right back into the new read_console hook function, over and over until it crashes.

**Comment by chrisant996** [comment]
If that's indeed what's going on, then it's a bit of a strange behavior from the anti-virus suite, but it may allow a potential workaround:

- Can you try `clink autorun show` and share the results?
- Is Clink set for autorun in both "Current user" and also "All users"?  If so, then Cmd will try to inject Clink twice every time.
- If so, then using `clink autorun uninstall` from a normal (_not_ administrator) command window would remove one of the inject calls, and then maybe the crashes might stop.

Or if you have a sequence of scripts and/or command line arguments that end up issuing more than one `clink inject` call in the same cmd.exe instance, then another workaround might be to remove the extras so there's only one inject attempt.

(Clink normally detects re-inject and ignores it; but if something interferes and makes Clink unable to detect that it's already injected, then Clink might accidentally inject itself again.)

**Comment by chrisant996** [comment]
Please do send the `clink.log` file.

But v1.2.32 includes a change that may help avoid this kind of issue, if it's a result of antivirus blocking certain parts of Clink's initialization, but permitting other parts of Clink's initialization.  Clink can't fully defend against interference or blocking by antivirus tools, but it can at least try to detect or accommodate certain kinds of non-fatal interference.

**Comment by tgice** [comment]
Hi Chris.  Thanks for the in-depth review.  Here's the latest clink.log running my 1.2.30.0ddffe version (which is still crashing).  I'll try upgrading too to see if that changes anything.

[clink.log](https://github.com/chrisant996/clink/files/7164230/clink.log)

As far as the autorun questions, I think I've always just had it set to run under 'Current user'.  E.g., autorun show looks like:

```
C:\WINDOWS\system32>clink autorun show
Current AutoRun values

  Current user:

    native : "C:\Program Files (x86)\clink\1.2.30.0ddffe\clink.bat" inject --autorun
     wow64 : "C:\Program Files (x86)\clink\1.2.30.0ddffe\clink.bat" inject --autorun

  All users:

    native : <unset>
     wow64 : <unset>
```

And trying to install or uninstall under an Admininstrator prompt doesn't seem to make any difference -- it always puts the changes in Current user -- ah, of course unless I use the '-a' option to switch to All users.

I tried that, having autorun set on both, and it doesn't seem to change the basic behavior: launch a new window, try `set /p a=`, have the window crash, so I'm guessing there's nothing to find here, unless you can think of something else specific to try (though it looks like I probably never had both Current & All users defined with autorun).

**Comment by tgice** [comment]
Updated to 1.2.32.  Still crashes at `set /p a=`.  Attached is a new clink.log.  Looks like nothing materially different in there.
[clink2.log](https://github.com/chrisant996/clink/files/7164264/clink2.log)

I just emailed you a new Time Travel trace or dump using 1.2.32 -- hopefully that's more helpful.

**Comment by chrisant996** [comment]
Thank you for the dumps and the Time Travel Trace!

The Time Travel Trace shows what is happening, but I don't understand yet why it's happening.  I'm still analyzing the trace.

Clink retrieves the address of one of its Import Address Table entries.  But it somehow actually gets one from a completely different DLL in memory.  And so when it goes to repair its IAT after hooking ReadConsoleW ... it "repairs" the entry in a different DLL, which already held the desired value.  Which leaves its own ReadConsoleW pointer not repaired, and so when it tries to call the original ReadConsoleW it gets the not-repaired address and ends up in infinite recursion, which exhausts the stack, and thus the stack overflow exception.

There is something strange going on with the IMAGE_IMPORT_DESCRIPTOR table.  I'll keep analyzing the trace over the next few evenings until I can figure out why this is happening.

**Would you be willing to capture one more Time Travel Trace (it was perfectly done, thank you!) if I send you a special build of Clink with some additional diagnostics?**  What's making it slow to debug is the compiler optimizations, do I'm doing everything in assembly language by hand.  And there is a lot of data structure offset arithmetic and dereferencing to follow the IMAGE_IMPORT_DESCRIPTOR tables.  It could go a lot faster if I send you a version with optimizations turned off (then I can debug via C++ source code instead of assembly language, and the debugger can automatically show the contents of variables) and also I can add additional log file output.

**Comment by chrisant996** [comment]
I manually analyzed the whole Import Address Table for clink_dll_x64.dll in the time travel trace.

It is simply not valid by the time the OS loader finishes loading the clink_dll_x64.dll.  Its format violates the spec for how the IAT works.  However, it is invalid in a way that will work ok as long as no software installs IAT hooks.  Maybe TrendMicro is intentionally redirecting the kernel32 IAT in order to interfere with malware?  I don't know.

It's theoretically possible to work around the read_console stack overflow.
But I verified in the Time Travel Trace that the other IAT hooks are also all broken.
And there isn't a way to work around those ones.

So I don't see any way to produce a workaround.  😢

**Comment by tgice** [comment]
Ugh!  That assembly debugging sounds terrible!  Thanks for plowing through all of it anyway.

I assume based on your second post that it won't do any good to give you another trace w/ a new special build then?

Also, you mentioned that it might be possible to workaround the read_console SO, but not the other IAT hooks -- what do those other ones involve?  The only crash I've run into so far is this prompt-based one, so I'm just curious as to what other features might not work or cause a crash based on your research.

Thanks again for all of your work on this.

Also, it's probably going to take my IT department a bit longer (maybe into next week) to take a closer look at what they're willing to do for me, but I'm really hoping we can at least agree on some sort of whitelisting for Clink.

**Comment by chrisant996** [comment]
More bad news:

I implemented the workaround. I completely rewrote the API hooking strategy. It seems to work ok for 64 bit.

But it can't even compile for 32 bit, because the linker names the hidden import library symbols with a character that isn't valid in C/C++ symbols, so there's no way to trick it into referencing them.

I can try to make additional changes to work around that, but it's becoming pretty invasive to the source code. The really concerning part is it would make it _very_ easy to accidentally break Clink in the future.

All this in an attempt to circumvent hidden corruption to core OS data structures. I'm not sure I _should_ be trying to make these changes. 😜

**Comment by tgice** [comment]
> All this in an attempt to circumvent hidden corruption to core OS data structures. I'm not sure I _should_ be trying to make these changes. 😜

Yeah, I hear you (I'm also a software developer, so I get it) ... I'd say don't do it!  Let me see where I can get with my IT department (a manager's been unavailable this week but probably will be next week) and hopefully that will wrap it up.

I'll report back when I get somewhere.

I really appreciate all of your efforts!

**Comment by chrisant996** [comment]
> I can try to make additional changes to work around that, but it's becoming pretty invasive to the source code. The really concerning part is it would make it _very_ easy to accidentally break Clink in the future.

Well, I kept going last night, and applied the additional workarounds.

And I found a spot where Clink's current IAT-based hooking strategy had a hidden bug.  So making the changes makes it easy to break Clink in one way, and _not_ making the changes makes it easy to break Clink in the opposite way.

And making the changes at least makes it more clear what's happening.

So I think I will probably proceed with the hooking rewrite.  I will do a bit more testing and validation, though.

**Comment by tgice** [comment]
Hi Chris, I will test it, but I've got a problem.  I'd noticed this before when I tried to install the portable version of clink -- Trend Micro just stepped all over it, kept popping these warning dialogs every time clink_x64 tried to run, and even when I got the 'autorun install' part done, it wouldn't actually start.

Here's a screenshot of the TM log for one of the reasons it's blocking clink, but I doubt that'll give you much insight into why it's happening (or maybe it does, but there's nothing you can do to work around it -- I'm sure TM is just naturally suspicious of software doing what clink is [legitimately] trying to do):

![image](https://user-images.githubusercontent.com/11559535/133822712-c366ca37-fc29-45a6-b477-77ed3e93397a.png)

Anyway, the way that I got the previous versions running (by trial and error) was _installing with the exe installer_.  Could you package this new version into an installer for me?  I still get the popped warnings, but for some reason once it's installed like that, the warnings stop.

Thanks.

**Comment by chrisant996** [comment]
GitHub does not allow linking .exe files in posts (and it blocks my attempt).

There is no more WriteVirtualMemory usage in Clink.  So trying the .zip version might work now.

Oh -- and I think I know why it works via the setup.exe program.
I think you need to right click on the downloaded .zip file, and choose Properties, and then choose Unblock and then OK.
And _then_ extract the files from the .zip file.  Trend Micro is probably applying more aggressive protection to files downloaded from the internet, until you Unblock them.

Here is an updated .zip:  
[clink.1.2.33.fd526c.zip](https://github.com/chrisant996/clink/files/7187178/clink.1.2.33.fd526c.zip)

**Comment by tgice** [comment]
Hey, I had thought of the unblock issue (of which I was aware) and had tried unblocking the individual files in the folder to which I'd unzipped, but that wasn't working.

So I did try your idea of unblocking the zip first (which I confirmed resulted in all of the internal files being unblocked automatically), but I'm still getting the same behavior: I get the blocked warning dialogs from Trend Micro Apex One each time I try running clink.bat.  Finally, once I do `autorun install`, and launch the next fresh cmd window, I get just the regular Windows copyright, no Clink one after that, and there's no clink functionality in the resulting Window (plus one more popup of TM I think, to which I answer 'allow once').

Could you email me a copy of the installer so I can try that, since I'm pretty sure it'll work around this TM issue of disallowing basic running from a portable install (for some reason I've still not figured out)?

**Comment by chrisant996** [comment]
I've put the setup exe inside a zip file.  Hopefully that lets you access it.
[clink.1.2.33.3b5485_setup.zip](https://github.com/chrisant996/clink/files/7188362/clink.1.2.33.3b5485_setup.zip)

**Comment by tgice** [comment]
Woohoo!  That did it!  `set /p a=` at least no longer crashes Clink on my PC with that troublesome Trend Micro running.

Do you think the changes are solid enough to leave in your main branch for now?

**Comment by chrisant996** [comment]
> Woohoo! That did it!

Thanks for the confirmation.  I'm still a bit creeped out that Trend Micro seems to be actively corrupting OS data structures, possibly intentionally, and there's nothing to say Trend Micro won't cause other interference with the Detours approach to hooking.  If so, then there's nothing more I can do.

> Do you think the changes are solid enough to leave in your main branch for now?

I am trying to do thorough testing before releasing an official update.  It's a very destabilizing change, and I've found and fixed 2 regressions.  They were glorified typos, and don't invalidate the approach.

It's too soon to say for sure, but I think it will be ok.

Since the changes are in the main branch already, I will close this issue for now.

**Comment by chrisant996** [comment]
@tgice, Detours hooking causes other problems, so I had to revert to IAT hooking.  While investigating #169, I realized Clink really does need to use IAT hooking in order for everything to work properly.

However, Detours seemed to circumvent Trend Micro's problem enough to get Clink generally working.  So I've added an **_unsupported_** `clink inject --detours` flag that forces using Detours hooking.  No one should use that flag unless it's absolutely necessary to work around interference from other programs.  And it's an unsupported flag.

If you still need Detours hooking, you'll need to use the --detours flag whenever injecting Clink (or in `clink autorun install -- --detours` if you want it installed for AutoRun).

**Comment by tgice** [comment]
Thanks Chris, I really wish that everyone would support software as well as you do (whether they're compensated or not!).

I appreciate you retaining these changes, and using the extra flag will be no problem for me as long as I need that to work around TrendMicro.
